dev/tests/integration/testsuite/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserverTest.php

```diff
@@ -210,25 +223,410 @@ public function testUrlKeyHasChangedInStoreviewContextWithoutPermanentRedirectio
+            $this->assertContains($row, $actual);
+        }
+    }
+
+    /**
+     * @magentoDataFixture Magento/Store/_files/second_website_with_two_stores.php
+     * @magentoDataFixture Magento/CatalogUrlRewrite/_files/product_rewrite_multistore.php
+     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
+     */
+    public function testChangeVisibilityGlobalScope()
+    {
+        $testStore1 = $this->storeManager->getStore('default');
+        $testStore2 = $this->storeManager->getStore('fixture_second_store');
+        $testStore3 = $this->storeManager->getStore('fixture_third_store');
+        $testStore4 = $this->storeManager->getStore('test');
+
+        $this->storeManager->setCurrentStore(Store::DEFAULT_STORE_ID);
+
+        /** @var Product $product*/
+        $product = $this->productRepository->get('product1');
+
+        $productFilter = [
+            UrlRewrite::ENTITY_TYPE => 'product',
+        ];
+
+        //Product in 1st website. Should result in being in 1st and 4th stores.
+        $expected = [
+            [
+                'request_path' => "product-1.html",
+                'target_path' => "catalog/product/view/id/" . $product->getId(),
+                'is_auto_generated' => 1,
+                'redirect_type' => 0,
+                'store_id' => $testStore1->getId(),
+            ],
+            [
+                'request_path' => "product-1.html",
+                'target_path' => "catalog/product/view/id/" . $product->getId(),
+                'is_auto_generated' => 1,
+                'redirect_type' => 0,
+                'store_id' => $testStore4->getId(),
+            ]
+        ];
+        $actual = $this->getActualResults($productFilter);
+        foreach ($expected as $row) {
+            $this->assertContains($row, $actual);
+        }
+
+        //Set product to be not visible at global scope
+        $this->storeManager->setCurrentStore(Store::DEFAULT_STORE_ID);
+        $product->setVisibility(Visibility::VISIBILITY_NOT_VISIBLE);
+        $this->productRepository->save($product);
+        $expected = [];
+        $actual = $this->getActualResults($productFilter);
+        foreach ($expected as $row) {
+            $this->assertContains($row, $actual);
+        }
+
+        //Add product to websites corresponding to all 4 stores.
+        //Rewrites should not be present as the product is hidden
+        //at the global scope.
+        $this->storeManager->setCurrentStore(Store::DEFAULT_STORE_ID);
+        $product->setWebsiteIds(
+            array_unique(
+                [
+                    $testStore1->getWebsiteId(),
+                    $testStore2->getWebsiteId(),
+                    $testStore3->getWebsiteId(),
+                    $testStore4->getWebsiteId()
+                ]
+            )
+        );
+        $this->productRepository->save($product);
+        $expected = [];
+        $actual = $this->getActualResults($productFilter);
+        foreach ($expected as $row) {
```

Comment:
Why do you need this if $expected is actually empty?

Reply:
changed asseration